File: chapter-01/binarysearch/binary_search.go
package binarysearch

import (
	"golang.org/x/exp/constraints"
)

func SearchIndex[T constraints.Ordered](array []T, element T) int {
	startIndex := 0
	endIndex := len(array) - 1

	for startIndex <= endIndex {
		middleIndex := (startIndex + endIndex) / 2
		guess := array[middleIndex]

		if guess == element {
			return middleIndex
		}

		if guess > element {
			endIndex = middleIndex - 1
		} else {
			startIndex = middleIndex + 1
		}
	}
	return -1
}

File: chapter-01/binary-search.py
def binary_search(nums: list[int], x: int) -> int:
    l, r = 0, len(nums) - 1
    while l <= r:
        mid = (l + r) // 2
        guess = nums[mid]
        if guess == x:
            return mid
        if guess < x:
            l = mid + 1
        else:
            r = mid - 1
    return -1

print(binary_search([1, 2, 3, 4, 5, 6], 5))